pkg/router/seller.go

```diff
@@ -1,269 +1,783 @@
-// @Description Add tag for shop.
-// @Tags Seller, Shop, Tag
-// @Accept json
-// @Produce json
-// @Success 200
-// @Failure 401
-// @Router /seller/tag [post]
+//	@Summary		Seller add tag
+//	@Description	Add tag for shop.
+//	@Tags			Seller, Shop, Tag
+//	@Accept			json
+//	@Param			name	body	string	true	"insert tag"	minlength(1)
+//	@Produce		json
+//	@success		200	{object}	db.SellerInsertTagRow
+//	@Failure		400	{object}	echo.HTTPError
+//	@Failure		409	{object}	echo.HTTPError
+//	@Failure		500	{object}	echo.HTTPError
+//	@Router			/seller/tag [post]
 func sellerAddTag(pg *db.DB, logger *zap.SugaredLogger) echo.HandlerFunc {
 	return func(c echo.Context) error {
-
-		return c.NoContent(http.StatusOK)
+		var username string = "user1"
+
+		var param db.HaveTagNameParams
+		if err := c.Bind(&param); err != nil {
+			logger.Error(err)
+			return err
+		}
+		if param.Name == "" || hasSpecialChars(param.Name) {
+			logger.Error("tag name invaild")
+			return echo.NewHTTPError(http.StatusInternalServerError, "tag name invaild")
+		}
+		param.SellerName = username
+		have, err := pg.Queries.HaveTagName(context.Background(), param)
+		if err != nil {
+			logger.Error(err)
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to check tag uniquity")
+		}
+		if have {
+			logger.Error("Conflict (tag name have to be unique")
+			return echo.NewHTTPError(http.StatusConflict, "Conflict (tag name have to be unique")
+		}
+		tag, err := pg.Queries.SellerInsertTag(context.Background(), db.SellerInsertTagParams(param))
+		if err != nil {
+			logger.Error(err)
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to add tag")
+		}
+		return c.JSON(http.StatusOK, tag)
 	}
 }
 
-// @Summary Seller get shop coupon
-// @Description Get all coupons for shop.
-// @Tags Seller, Shop, Coupon
-// @Produce json
-// @Success 200
-// @Failure 401
-// @Router /seller/coupon [get]
+//	@Summary		Seller get shop coupon
+//	@Description	Get all coupons for shop.
+//	@Tags			Seller, Shop, Coupon
+//	@Param			offset	body	int	true	"offset page"	minimum(0)
+//	@Produce		json
+//	@success		200	{object}	[]db.SellerGetCouponRow
+//	@Failure		400	{object}	echo.HTTPError
+//	@Failure		500	{object}	echo.HTTPError
+//	@Router			/seller/coupon [get]
 func sellerGetShopCoupon(pg *db.DB, logger *zap.SugaredLogger) echo.HandlerFunc {
 	return func(c echo.Context) error {
-
-		return c.NoContent(http.StatusOK)
+		var username string = "user1"
+		var couponPerPage int32 = 20
+
+		var param db.SellerGetCouponParams
+		if err := c.Bind(&param); err != nil {
+			logger.Error(err)
+			return err
+		}
+		param.SellerName = username
+		param.Limit = couponPerPage
+		param.Offset = param.Offset * couponPerPage
+		coupons, err := pg.Queries.SellerGetCoupon(context.Background(), param)
+		if err != nil {
+			logger.Error(err)
+			return echo.NewHTTPError(http.StatusInternalServerError, "failed to get seller coupon")
+		}
+		return c.JSON(http.StatusOK, coupons)
 	}
 }
```

Comment:
Consider both limit and offset as path parameter for dynamic layout 🪄
The defination of offset should equals to the raw offset, nor use somehing like page number.

Reply:
It seems like sqlc cannot add the go_struct_tag on a non-column variable, and I don't think it is good idea to let the front end decide how much data they want in one request, especially considering that the front end may not reliable.

Reply:
- You can create custom struct to get those query parameter. 
- The API user should be able to set limit, but have max limit at the same time.